File: backend/app/models/quiz.py
from pydantic import BaseModel
from typing import List
from datetime import datetime

class QuizOption(BaseModel):
    text: str
    is_correct: bool

class QuizQuestion(BaseModel):
    question: str
    options: List[str]  
    correct_answer: str
    explanation: str

class QuizGenerateRequest(BaseModel):
    video_id: str
    num_questions: int = 5

class QuizResponse(BaseModel):
    questions: List[QuizQuestion]
    video_id: str
    generated_at: datetime

class QuizAnswer(BaseModel):
    question_index: int
    selected_answer: str

class QuizSubmission(BaseModel):
    video_id: str
    answers: List[QuizAnswer]

class QuizResult(BaseModel):
    video_id: str
    score: int
    total_questions: int
    correct_answers: List[int]  # indices of correct answers
    submitted_at: datetime
    time_taken: int  # seconds

class QuizResultResponse(BaseModel):
    result: QuizResult
    questions: List[QuizQuestion]  # Include questions for review

File: frontend/mercurious/src/lib/api/types/quiz.ts
export interface QuizQuestion {
  question: string;
  options: string[];
  correct_answer: string;
  explanation: string;
}

export interface QuizGenerateRequest {
  video_id: string;
  num_questions?: number;
}

export interface QuizResponse {
  questions: QuizQuestion[];
  video_id: string;
  generated_at: string;
}

export interface QuizAnswer {
  question_index: number;
  selected_answer: string;
}

export interface QuizSubmission {
  video_id: string;
  answers: QuizAnswer[];
}

export interface QuizResult {
  video_id: string;
  score: number;
  total_questions: number;
  correct_answers: number[];
  submitted_at: string;
  time_taken: number;
}

export interface QuizResultResponse {
  result: QuizResult;
  questions: QuizQuestion[];
}

export interface QuizStatistics {
  total_videos_with_quizzes: number;
  total_quiz_attempts: number;
  overall_average_score: number;
  best_overall_score: number;
  completion_rate: number;
}

export interface QuizAvailability {
  video_id: string;
  quiz_available: boolean;
  quiz_fresh: boolean;
  generated_at: string | null;
}

// UI-specific types for better user experience
export interface QuizState {
  currentQuestionIndex: number;
  answers: QuizAnswer[];
  timeStarted: Date | null;
  timeElapsed: number;
  isSubmitted: boolean;
  result: QuizResultResponse | null;
}

export interface QuizProgress {
  currentQuestion: number;
  totalQuestions: number;
  answeredQuestions: number;
  percentComplete: number;
}

// Error types for better error handling
export interface QuizError {
  type: 'generation' | 'submission' | 'loading' | 'network';
  message: string;
  details?: string;
}
